stl/inc/xutility

```diff
@@ -4708,7 +4704,7 @@ _NODISCARD _CONSTEXPR20 bool equal(const _InIt1 _First1, const _InIt1 _Last1, co
     auto _UFirst1      = _Get_unwrapped(_First1);
     const auto _ULast1 = _Get_unwrapped(_Last1);
     auto _UFirst2      = _Get_unwrapped_n(_First2, _Idl_distance<_InIt1>(_UFirst1, _ULast1));
-    if constexpr (decltype(_Equal_memcmp_is_safe(_UFirst1, _UFirst2, _Pred))::value) {
+    if constexpr (_Equal_memcmp_is_safe<_InIt1, _InIt2, _Pr>) {

```

Comment:
This transformation is incorrect, preventing the optimization from activating in certain situations. "Unwrapping" iterators (e.g. from `vector<T>::iterator` to `T*`) serves two purposes:

* it increases performance by front-loading iterator debugging checks followed by direct use of raw pointers, and
* it makes metaprogramming optimizations easier.

Here, `_Equal_memcmp_is_safe` recognizes raw pointers, and then inspects the elements to see whether they're eligible for `memcmp`. It doesn't know how to recognize `vector::iterator`, but that's fine, because `vector::iterator` knows how to unwrap to a raw pointer.

Here, we were previously passing the unwrapped `_UFirst1, _UFirst2`, inspecting the unwrapped iterators. Now, you're passing the original `_InIt1, _InIt2`, which are still wrapped. This will activate the optimization only if the user actually passed raw pointers, not if they passed `vector::iterator` etc.

Reply:
I am absolutely amazed that you found that bug. 

I changed it to `_Equal_memcmp_is_safe<decltype(_UFirst), decltype(_UFirst2), _Pr>`

that should pass the unwrapped raw pointer to `_Equal_memcmp_is_safe`